### PR TITLE
Lint that newly added components use the new folder structure

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

* Add an allowlist of existing components that use the classic component structure
* Add source tree in testdata that defines classic and new-style components
* Add a basic unit test for get_components_and_bundles, which many lint rules depend on
* Add TODOs for other things the linter should be checking in the new component layout
* Remove some components_to_migrate that currently pass the existing linters

### Motivation

Enforcing consistency with the new components layout: https://datadoghq.dev/datadog-agent/components/creating-components/

### CODE REVIEW

File: tasks/components.py [hunk 1]
```diff
@@ -102,9 +184,17 @@ def check_component_contents_and_file_hiearchy(entry_point):
         return ""
 
     for folder in directory.iterdir():
+        # TODO: Check entry_point.version == 2 for new-style, and entry_point.version == 1 for classic (currently broken on some classic components)
+
+        # Check for component implementation using the new-style folder structure: comp/<comopnent>/impl[-suffix]
```
Comment (ogaca-dd) [inline tasks/components.py:189]:
```suggestion
        # Check for component implementation using the new-style folder structure: comp/<component>/impl[-suffix]
```

**Review by ogaca-dd — APPROVED** [review]
LGTM

**Review by ogaca-dd — APPROVED** [review]

File: tasks/gotest.py [hunk 2]
```diff
@@ -838,6 +833,16 @@ def should_run_all_tests(files, trigger_files):
     return False
 
 
+def get_go_modified_files(ctx):
+    files = get_modified_files(ctx)
+    return [
+        file
+        for file in files
+        if file.find("unit-tests/testdata/components_src") == -1
```
Comment (GustavoCaso) [inline tasks/gotest.py:841]: We ensure that we so not try running tests for these dummy go files 😄

File: tasks/components.py [hunk 3]
```diff
@@ -201,9 +307,15 @@ def locate_root(dir):
     # v2 component: this folder is a component root if it contains 'def/component.go'
     component_file = dir / 'def/component.go'
     if component_file.is_file():
-        return ComponentRoot(component_file, dir, 2)
+        # comp/api/api/def/component.go is a special case, it's not a component using version 2
+        # PLEASE DO NOT ADD MORE EXCPETIONS
```
Comment (dustmop) [inline tasks/components.py:311]:
```suggestion
        # PLEASE DO NOT ADD MORE EXCEPTIONS
```

File: tasks/unit-tests/components_tests.py [hunk 4]
```diff
@@ -8,13 +8,8 @@
 
 class TestComponents(unittest.TestCase):
     def setUp(self):
-        # Add the classic component (used by tests) to the allowlist
-        classicComp = 'comp/classic/classicimpl'
-        if classicComp not in components.components_classic_style:
-            components.components_classic_style.append(classicComp)
         # Create a temporary directory of the source directory to allow
         # tests to modify the source files
-        # TODO: Add tests that modify source files to ensure that errors work
```
Comment (dustmop) [inline tasks/unit-tests/components_tests.py]: Could you add this TODO back, please? I think it would be useful to have more fine-grain tests. For example, since this test is using `shutil.copytree` to copy the test folder to tmp, we can modify the `new_style` folder to break in various ways:
- Have it use `impl_suffix` instead of `impl-suffix`
- Call the file `def/interface.go` instead of `def/component.go`
- Use a package name for the impl folder that doesn't match `package <component>impl`.
- etc
Basically this could be a list of checks in the function `check_component_contents_and_file_hiearchy` and we could use this technique to get very high coverage of all of the rules we want to enforce. In other words, I don't expect this TODO to be resolved in this PR, as it's only introducing the single rule of "don't use the classic style" going forward, but there's much more the linter can do.

**Review by dustmop — COMMENTED** [review]
The commits applied on top of my initial work mostly look good, with two small requests for fixes.

File: tasks/unit-tests/components_tests.py (hunk 4 above)
Comment (dustmop) [reply tasks/unit-tests/components_tests.py]: Adding it back in with my own commit

File: tasks/unit-tests/components_tests.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+import os
```
Comment (pducolin) [inline tasks/unit-tests/components_tests.py:1]: **💬 suggestion**
Can you take ownership of this file and of `test-data/components_src`?

**Review by CelianR — APPROVED** [review]

File: tasks/unit-tests/components_tests.py (hunk 5 above)
Comment (dustmop) [reply tasks/unit-tests/components_tests.py:1]: Done.

File: tasks/unit-tests/components_tests.py (hunk 4 above)
Comment (GustavoCaso) [reply tasks/unit-tests/components_tests.py]: Thank you so much for the explanation. 
My reasoning to remove the TODO was because I added a test that check that `get_components_and_bundles` return errors if the component is not path of the exception. But you are absolutely right, that we need to validate issues with components that are not part of the exception lists 😄

**Review by GustavoCaso — APPROVED** [review]

File: .golangci.yml [hunk 6]
```diff
@@ -14,7 +14,7 @@ issues:
 
   # Do not limit the number of times a same issue is reported.
   max-same-issues: 0
-
+
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply .golangci.yml:17]: That was my bad 😄

[the remark this answers is not in the document]
Comment (dustmop) [reply .golangci.yml:17]: I can remove it, also I need to resolve the codeowners conflict.
